src/Microsoft.Health.Fhir.R4.Core/Features/Resources/Upsert/UpsertResourceHandler.cs

```diff
@@ -0,0 +1,24 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Messages.Upsert;
+
+namespace Microsoft.Health.Fhir.Core.Features.Resources.Upsert
+{
+    public partial class UpsertResourceHandler : BaseResourceHandler, IRequestHandler<UpsertResourceRequest, UpsertResourceResponse>
+    {
+        private async Task<UpsertOutcome> UpsertAsync(UpsertResourceRequest message, ResourceWrapper resourceWrapper, bool allowCreate, bool keepHistory, CancellationToken cancellationToken)
+        {
+            // In R4, the server returns a 412 Precondition Failed status code if the version id given in the If-Match header does not match
```

Comment:
Is this for R4 only?

http://hl7.org/fhir/STU3/http.html#concurrency seems to indicate that it should also return 412.

Reply:
I'm fairly confident it's just for R4 - in the section that you linked to in the Stu3 spec, there's a line that says:

> If the version id given in the If-Match header does not match, the server returns a 409 Conflict status code instead of updating the resource

Reply:
oh you are right. I missed that part.